back/sssscs/src/main/java/com/ib/certificate/CertificateService.java

```diff
@@ -25,11 +57,182 @@ public Certificate findById(Long id) {
+		Certificate parent = c.getParent();
+		X509Certificate parent509 = keyUtil.getX509Certificate(parent.getSerialNumber());
+		String issuerName = parent509.getSubjectX500Principal().getName(); // TODO: Formatting. Or some other way? Same for endpoint where we're getting certificates' summaries.
+		
+		try {	
+			JcaContentSignerBuilder csbuilder = new JcaContentSignerBuilder("SHA256WithRSAEncryption");
+			csbuilder = csbuilder.setProvider("BC");
+			ContentSigner signer = csbuilder.build(issuerKey);
+			X509v3CertificateBuilder builder = new JcaX509v3CertificateBuilder(
+				keyUtil.getX500Name(issuerName),
+				BigInteger.valueOf(c.getId()),
+				keyUtil.dateFrom(validFrom),
+				keyUtil.dateFrom(validTo),
+				keyUtil.getX500Name(req.getSubjectName()),
+				subjectKey
+			);
+			
+			X509CertificateHolder holder = builder.build(signer);
+			JcaX509CertificateConverter conv = new JcaX509CertificateConverter();
+			conv = conv.setProvider("BC");
+			return conv.getCertificate(holder);
+		} catch (Exception e) {
+			e.printStackTrace();
+			return null;
+		}
+	}
+
+	private X509Certificate generateSelfSigned(KeyPair kp, LocalDateTime validFrom, LocalDateTime validTo) {
+		try {	
+			
+			X500Name x500Name = new X500Name("CN=localhost");
+			JcaContentSignerBuilder csbuilder = new JcaContentSignerBuilder("SHA256WithRSAEncryption");
+			csbuilder = csbuilder.setProvider("BC");
+			ContentSigner signer = csbuilder.build(kp.getPrivate());
+			X509v3CertificateBuilder builder = new JcaX509v3CertificateBuilder(
+				x500Name,
+				BigInteger.valueOf(Instant.now().getEpochSecond()),
+				keyUtil.dateFrom(validFrom),
+				keyUtil.dateFrom(validTo),
+				x500Name,
+				kp.getPublic()
+			);
+			
+			X509CertificateHolder holder = builder.build(signer);
+			JcaX509CertificateConverter conv = new JcaX509CertificateConverter();
+			conv = conv.setProvider("BC");
+			return conv.getCertificate(holder);
+		} catch (Exception e) {
+			e.printStackTrace();
+			return null;
+		}
+	}
+		
 	@Override
 	public List<Certificate> getAll() {
 		return certificateRepo.findAll();
 	}
+
+	@Override
+	public boolean validate(Certificate cert) {
+		if (isExpired(cert)) return false;
+		if (isInvalidSignature(cert)) return false;
+		
+		if (isTrusted(cert)) {
+			return true;
+		} else {			
+			return validate(cert.getParent());
+		}
+	}
+
+	private boolean isTrusted(Certificate cert) {
+		return (cert.getType() == Type.ROOT);
+	}
+	
+	private boolean isExpired(Certificate cert) {
+		if (cert.getValidFrom().isAfter(LocalDateTime.now())) {
+			return true;
+		}
+		if (LocalDateTime.now().isAfter(cert.getValidTo())) {
+			return true;
+		}
+		return false;
+		//return (LocalDateTime.now().isBefore(cert.getValidFrom()) || LocalDateTime.now().isAfter(cert.getValidTo()));
+	}
+	
+	private boolean isInvalidSignature(Certificate cert) {
+		if (isTrusted(cert)) {
+			return false;
+		}
```

Comment:
Don't you still need to check if it is signed with the private key?? IsTrusted only checks if it's root, right?

Reply:
You're right, fixed in [695454c](https://github.com/magley/sssscs/pull/5/commits/695454cfc60f54726397c5cc6dbc1f3638123d75).